### ISSUE TITLE
Rate limit has been exceeded for: testflight|default|APPLE_CONNECT

### ISSUE THREAD

**Issue by elviin** [fastlane/fastlane#12845]
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description
From time to time I get an error: Rate limit has been exceeded for: testflight|default|APPLE_CONNECT:11807886017

 But the application is successfully uploaded to Testflight anyway. 

My apple account is used for several appstoreconnect teams and this issues happens almost randomly with different CircleCI builds.

##### Complete output when running fastlane, including the stack trace and command used

<details>
  <pre>DEBUG [2018-07-03 06:50:18.12]: [Transporter]: INFO:      File: f00975d02270cbf57a185385f8550a91d5d0d37521b474d76a78df1ea788bd39.ipa 88559000/106680968, 83% completed

DEBUG [2018-07-03 06:50:18.29]: [Transporter]: INFO:      File: f00975d02270cbf57a185385f8550a91d5d0d37521b474d76a78df1ea788bd39.ipa 90692000/106680968, 85% completed

DEBUG [2018-07-03 06:50:18.36]: [Transporter]: INFO:      File: f00975d02270cbf57a185385f8550a91d5d0d37521b474d76a78df1ea788bd39.ipa 91750600/106680968, 86% completed

DEBUG [2018-07-03 06:50:18.44]: [Transporter]: INFO:      File: f00975d02270cbf57a185385f8550a91d5d0d37521b474d76a78df1ea788bd39.ipa 92825000/106680968, 87% completed

DEBUG [2018-07-03 06:50:18.59]: [Transporter]: INFO:      File: f00975d02270cbf57a185385f8550a91d5d0d37521b474d76a78df1ea788bd39.ipa 94958000/106680968, 89% completed

DEBUG [2018-07-03 06:50:18.75]: [Transporter]: INFO:      File: f00975d02270cbf57a185385f8550a91d5d0d37521b474d76a78df1ea788bd39.ipa 97091000/106680968, 91% completed

DEBUG [2018-07-03 06:50:18.90]: [Transporter]: INFO:      File: f00975d02270cbf57a185385f8550a91d5d0d37521b474d76a78df1ea788bd39.ipa 98149600/106680968, 92% completed

DEBUG [2018-07-03 06:50:18.98]: [Transporter]: INFO:      File: f00975d02270cbf57a185385f8550a91d5d0d37521b474d76a78df1ea788bd39.ipa 99224000/106680968, 93% completed

DEBUG [2018-07-03 06:50:19.17]: [Transporter]: INFO:      File: f00975d02270cbf57a185385f8550a91d5d0d37521b474d76a78df1ea788bd39.ipa 101357000/106680968, 95% completed

DEBUG [2018-07-03 06:50:19.33]: [Transporter]: INFO:      File: f00975d02270cbf57a185385f8550a91d5d0d37521b474d76a78df1ea788bd39.ipa 103490000/106680968, 97% completed

DEBUG [2018-07-03 06:50:19.41]: [Transporter]: INFO:      File: f00975d02270cbf57a185385f8550a91d5d0d37521b474d76a78df1ea788bd39.ipa 104548600/106680968, 98% completed

DEBUG [2018-07-03 06:50:19.52]: [Transporter]: INFO:      File: f00975d02270cbf57a185385f8550a91d5d0d37521b474d76a78df1ea788bd39.ipa 105623000/106680968, 99% completed

DEBUG [2018-07-03 06:50:19.60]: [Transporter]: INFO:      File: f00975d02270cbf57a185385f8550a91d5d0d37521b474d76a78df1ea788bd39.ipa 106680968/106680968, 100% completed

DEBUG [2018-07-03 06:50:19.70]: [Transporter]: DBG-X: The status for the file, f00975d02270cbf57a185385f8550a91d5d0d37521b474d76a78df1ea788bd39.ipa, is TRANSFERRED.

DEBUG [2018-07-03 06:50:19.70]: [Transporter]: INFO: Package upload progress: 100% completed

DEBUG [2018-07-03 06:50:20.02]: [Transporter]: INFO: The Signiant transfer engine's status is DISCONNECTED

DEBUG [2018-07-03 06:50:20.09]: [Transporter]: INFO: The Signiant transfer engine's status is DISCONNECTED

DEBUG [2018-07-03 06:50:20.14]: [Transporter]: INFO: The file, metadata.xml, was uploaded successfully.

DEBUG [2018-07-03 06:50:20.14]: [Transporter]: INFO: The file, f00975d02270cbf57a185385f8550a91d5d0d37521b474d76a78df1ea788bd39.ipa, was uploaded successfully.

DEBUG [2018-07-03 06:50:20.14]: [Transporter]: INFO: The status for the upload is SUCCESSFUL.

DEBUG [2018-07-03 06:50:20.14]: [Transporter]: INFO: The package's files (1363154774.itmsp) uploaded successfully.

DEBUG [2018-07-03 06:50:20.14]: [Transporter]: INFO: Done performing upload.

DEBUG [2018-07-03 06:50:20.14]: [Transporter]: DBG-X: The transfer time for the package 1363154774.itmsp was 11999 milliseconds.

DEBUG [2018-07-03 06:50:20.14]: [Transporter]: DBG-X: Throughput information for 1363154774.itmsp: 0 minutes, 12 seconds, 101.74MB, 8.48MB/sec

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X: Deflated 472 bytes to 348

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X: Memory: [JVM] 933M free, 981M total, 981M max [System] (Physical) 3524M free, 8192M total (Swap) 0 free, 0 total

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X: Using operation named: uploadDoneWithArguments

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X: Apple's web service operation input parameters:

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X:   parameter Application = iTMSTransporter

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X:   parameter BaseVersion = 1.9.6

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X:   parameter CPUToken = 0f278e32-341c-42a7-bab4-98d1dc8f4432/1530625796963

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X:   parameter ClientChecksumInfo = [{CalculationTime=343, FileLastModified=1530625718000, Filename=f00975d02270cbf57a185385f8550a91d5d0d37521b474d76a78df1ea788bd39.ipa, FileSize=106680968, CalculatedChecksum=91b33221f580a3577c22eaed411e51e6}]

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X:   parameter ExtraTransportLogCompressed = (suppressed)

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X:   parameter FileSizeInfo = {f00975d02270cbf57a185385f8550a91d5d0d37521b474d76a78df1ea788bd39.ipa=106680968, metadata.xml=485}

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X:   parameter NewPackageName = 1363154774-20180703064957-382.itmsp

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X:   parameter NumberBytesTransferred = 106681453

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X:   parameter OSIdentifier = Mac OS X 10.13.3 (x86_64); jvm=25.71-b02; jre=1.8.0-u131-b11-iTMSTransporter-b02

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X:   parameter PackagePathWithoutBase = sgr8_pod2/transporter/content/transporter/HC34URQPCY

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X:   parameter StatisticsArray = [{duration=0.301362675, method=clientApplication.start}, {duration=0.399900324, method=authenticateForSession}, {duration=3.717172344, method=validateMetadata}, {duration=9.109273543, method=lookupTransportDiagnostic}, {duration=24.261659574, method=validateAssets}, {duration=1.040129615, method=clientChecksumCompleted}]

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X:   parameter StatisticsClientStartDateTimeZoneISO = 2018-07-03T06:50:20-07:00

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X:   parameter StatisticsPreviousCallDurationInSecs = 1.040129615

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X:   parameter Throttle = 100000

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X:   parameter TransferTime = 11999

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X:   parameter Transport = Signiant

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X:   parameter TransportLogCompressed = (suppressed)

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X:   parameter TransporterArguments = -m upload -u vladimir.slavik@clark.de -p **hidden value** -f /var/folders/gk/lkr9pm5x039fx6d3j9r52rv80000gn/T/d20180703-634-1knsphe/1363154774.itmsp -t Signiant -k 100000

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X:   parameter Username = vladimir.slavik@clark.de

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X:   parameter Version = 1.9.6

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: DBG-X:   parameter iTMSTransporterMode = upload

DEBUG [2018-07-03 06:50:20.15]: [Transporter]: INFO: id = 20180703065020-053

DEBUG [2018-07-03 06:50:22.09]: [Transporter]: DBG-X: Apple's web service operation return value:

DEBUG [2018-07-03 06:50:22.09]: [Transporter]: DBG-X:   parameter EnableJWTForAllCalls = false

DEBUG [2018-07-03 06:50:22.09]: [Transporter]: DBG-X:   parameter SessionInvalid = true

DEBUG [2018-07-03 06:50:22.09]: [Transporter]: DBG-X:   parameter Success = true

DEBUG [2018-07-03 06:50:22.09]: [Transporter]: DBG-X:   parameter StreamingSettings = {LogStreamingEnabled=true, MultipartUploadsEnabled=true, AssetDescriptionStreamingEnabled=false}

DEBUG [2018-07-03 06:50:22.09]: [Transporter]: DBG-X: Unlocking token at path /Users/distiller/.itmstransporter/UploadTokens/static.162.252.208.84.cyberlynk.net_HC34URQPCY_1363154774.itmsp.token

DEBUG [2018-07-03 06:50:22.09]: [Transporter]: DBG-X: Releasing token file lock

DEBUG [2018-07-03 06:50:22.09]: [Transporter]: DBG-X: Removing token at path /Users/distiller/.itmstransporter/UploadTokens/static.162.252.208.84.cyberlynk.net_HC34URQPCY_1363154774.itmsp.token

DEBUG [2018-07-03 06:50:22.09]: [Transporter]: INFO: Done performing uploadDone notification to Apple.

DEBUG [2018-07-03 06:50:22.09]: [Transporter]: INFO: The package: /var/folders/gk/lkr9pm5x039fx6d3j9r52rv80000gn/T/d20180703-634-1knsphe/1363154774.itmsp has been successfully uploaded.

INFO [2018-07-03 06:50:22.22]: iTunes Transporter successfully finished its job
DEBUG [2018-07-03 06:50:22.22]: [Transporter]: DBG-X: Returning 0

WARN [2018-07-03 06:50:22.32]: [iTMSTransporter] [2018-07-03 06:50:22 PDT] <main> DBG-X:   parameter Success = true

WARN [2018-07-03 06:50:22.32]: [iTMSTransporter] [2018-07-03 06:50:22 PDT] <main> DBG-X:   parameter StreamingSettings = {LogStreamingEnabled=true, MultipartUploadsEnabled=true, AssetDescriptionStreamingEnabled=false}

WARN [2018-07-03 06:50:22.32]: [iTMSTransporter] [2018-07-03 06:50:22 PDT] <main> DBG-X: Unlocking token at path /Users/distiller/.itmstransporter/UploadTokens/static.162.252.208.84.cyberlynk.net_HC34URQPCY_1363154774.itmsp.token

WARN [2018-07-03 06:50:22.32]: [iTMSTransporter] [2018-07-03 06:50:22 PDT] <main> DBG-X: Releasing token file lock

WARN [2018-07-03 06:50:22.32]: [iTMSTransporter] [2018-07-03 06:50:22 PDT] <main> DBG-X: Removing token at path /Users/distiller/.itmstransporter/UploadTokens/static.162.252.208.84.cyberlynk.net_HC34URQPCY_1363154774.itmsp.token

WARN [2018-07-03 06:50:22.32]: [iTMSTransporter] [2018-07-03 06:50:22 PDT] <main>  INFO: Done performing uploadDone notification to Apple.

WARN [2018-07-03 06:50:22.32]: [iTMSTransporter] [2018-07-03 06:50:22 PDT] <main>  INFO: The package: /var/folders/gk/lkr9pm5x039fx6d3j9r52rv80000gn/T/d20180703-634-1knsphe/1363154774.itmsp has been successfully uploaded.

WARN [2018-07-03 06:50:22.32]: [iTMSTransporter] 

WARN [2018-07-03 06:50:22.32]: [iTMSTransporter] 

WARN [2018-07-03 06:50:22.32]: [iTMSTransporter] 

WARN [2018-07-03 06:50:22.32]: [iTMSTransporter] Package Summary:

WARN [2018-07-03 06:50:22.32]: [iTMSTransporter]  

WARN [2018-07-03 06:50:22.32]: [iTMSTransporter] 1 packages were uploaded successfully:

WARN [2018-07-03 06:50:22.32]: [iTMSTransporter] 	/var/folders/gk/lkr9pm5x039fx6d3j9r52rv80000gn/T/d20180703-634-1knsphe/1363154774.itmsp

WARN [2018-07-03 06:50:22.32]: [iTMSTransporter] [2018-07-03 06:50:22 PDT] <main> DBG-X: Returning 0

INFO [2018-07-03 06:50:22.32]: iTunes Transporter output above ^

WARN [2018-07-03 06:50:22.32]: Although errors occurred during execution of iTMSTransporter, it returned success status.
INFO [2018-07-03 06:50:22.36]: -----------------------------------------------------
INFO [2018-07-03 06:50:22.36]: Successfully uploaded package to iTunes Connect. It might take a few minutes until it's visible online.
INFO [2018-07-03 06:50:22.36]: -----------------------------------------------------
INFO [2018-07-03 06:50:22.37]: Successfully uploaded the new binary to iTunes Connect
INFO [2018-07-03 06:50:22.37]: If you want to skip waiting for the processing to be finished, use the `skip_waiting_for_build_processing` option
INFO [2018-07-03 06:50:31.02]: Waiting for iTunes Connect to finish processing the new build (2.0.3 - 3114.35425.0)
INFO [2018-07-03 06:51:03.46]: Waiting for iTunes Connect to finish processing the new build (2.0.3 - 3114.35425.0)
INFO [2018-07-03 06:51:36.20]: Waiting for iTunes Connect to finish processing the new build (2.0.3 - 3114.35425.0)
INFO [2018-07-03 06:52:09.51]: Waiting for iTunes Connect to finish processing the new build (2.0.3 - 3114.35425.0)
INFO [2018-07-03 06:52:42.30]: Waiting for iTunes Connect to finish processing the new build (2.0.3 - 3114.35425.0)
INFO [2018-07-03 06:53:14.92]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
INFO [2018-07-03 06:53:47.73]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
INFO [2018-07-03 06:54:20.31]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
INFO [2018-07-03 06:54:53.31]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
INFO [2018-07-03 06:55:26.02]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
INFO [2018-07-03 06:55:58.54]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
INFO [2018-07-03 06:56:31.05]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
INFO [2018-07-03 06:57:03.88]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
INFO [2018-07-03 06:57:37.08]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
INFO [2018-07-03 06:58:10.18]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
INFO [2018-07-03 06:58:43.08]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
INFO [2018-07-03 06:59:14.98]: Build doesn't show up in the build list anymore, waiting for it to appear again (check your email for processing issues if this continues)
WARN [2018-07-03 06:59:45.20]: Lane Context:
INFO [2018-07-03 06:59:45.20]: {:DEFAULT_PLATFORM=>:ios, :GET_INFO_PLIST_VALUE_CUSTOM_VALUE=>"2.0.3", :PLATFORM_NAME=>:ios, :LANE_NAME=>"ios beta_testflight", :ORIGINAL_DEFAULT_KEYCHAIN=>"\"/Users/distiller/Library/Keychains/login.keychain-db\"", :SIGH_PROFILE_TYPE=>"app-store", :MATCH_PROVISIONING_PROFILE_MAPPING=>{"de.devk.devk2go"=>"match AppStore de.devk.devk2go"}, :IPA_OUTPUT_PATH=>"/Users/distiller/project/DEVK.ipa", :DSYM_OUTPUT_PATH=>"/Users/distiller/project/DEVK.app.dSYM.zip", :XCODEBUILD_ARCHIVE=>"/Users/distiller/Library/Developer/Xcode/Archives/2018-07-03/DEVK 2018-07-03 06.46.36.xcarchive"}
ERROR [2018-07-03 06:59:45.20]: Rate limit has been exceeded for: testflight|default|APPLE_CONNECT:11807886017

INFO [2018-07-03 06:59:45.20]: Successfully generated documentation at path '/Users/distiller/project/fastlane/README.md'

+------+-----------------------------+-------------+
|                 fastlane summary                 |
+------+-----------------------------+-------------+
| Step | Action                      | Time (in s) |
+------+-----------------------------+-------------+
| 1    | Verifying fastlane version  | 0           |
| 2    | default_platform            | 0           |
| 3    | get_info_plist_value        | 0           |
| 4    | git log -1 --pretty=%B 2>&1 | 0           |
| 5    | fastlane -v                 | 2           |
| 6    | setup_circle_ci             | 0           |
| 7    | cocoapods                   | 30          |
| 8    | shell command               | 0           |
| 9    | update_info_plist           | 0           |
| 10   | match                       | 12          |
| 11   | gym                         | 120         |
| 💥   | pilot                       | 670         |
+------+-----------------------------+-------------+

DEBUG [2018-07-03 06:59:45.21]: All plugins are up to date
ERROR [2018-07-03 06:59:45.21]: fastlane finished with errors</pre>
</details>

### Environment

<details><summary>🚫 fastlane environment 🚫</summary>

### Stack

| Key                         | Value                                       |
| --------------------------- | ------------------------------------------- |
| OS                          | 10.13.5                                     |
| Ruby                        | 2.4.1                                       |
| Bundler?                    | true                                        |
| Git                         | git version 2.15.1                          |
| Installation Source         | ~/.rvm/gems/ruby-2.4.1/bin/fastlane         |
| Host                        | Mac OS X 10.13.5 (17F77)                    |
| Ruby Lib Dir                | ~/.rvm/rubies/ruby-2.4.1/lib                |
| OpenSSL Version             | OpenSSL 1.1.0g  2 Nov 2017                  |
| Is contained                | false                                       |
| Is homebrew                 | false                                       |
| Is installed via Fabric.app | false                                       |
| Xcode Path                  | /Applications/Xcode.app/Contents/Developer/ |
| Xcode Version               | 9.4                                         |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_US.UTF-8 | ✅ |
| LC_ALL   | en_US.UTF-8 | ✅ |
| LANGUAGE |             |   |

### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

<details>
  <pre>### fastlane gems

| Gem      | Version | Update-Status      |
| -------- | ------- | ------------------ |
| fastlane | 2.98.0  | 🚫 Update available |

### Loaded fastlane plugins:

| Plugin                 | Version | Update-Status |
| ---------------------- | ------- | ------------- |
| fastlane-plugin-sentry | 1.5.0   | ✅ Up-To-Date  |

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| did_you_mean              | 1.1.0        |
| executable-hooks          | 1.4.2        |
| bundler-unload            | 1.0.2        |
| rubygems-bundler          | 1.4.4        |
| bundler                   | 1.16.1       |
| io-console                | 0.4.6        |
| CFPropertyList            | 3.0.0        |
| concurrent-ruby           | 1.0.5        |
| i18n                      | 0.9.1        |
| minitest                  | 5.10.3       |
| thread_safe               | 0.3.6        |
| tzinfo                    | 1.2.4        |
| activesupport             | 4.2.10       |
| public_suffix             | 2.0.5        |
| addressable               | 2.5.2        |
| atomos                    | 0.1.2        |
| babosa                    | 1.0.2        |
| claide                    | 1.0.2        |
| fuzzy_match               | 2.0.4        |
| nap                       | 1.1.0        |
| cocoapods-core            | 1.3.1        |
| cocoapods-deintegrate     | 1.0.1        |
| cocoapods-downloader      | 1.1.3        |
| cocoapods-plugins         | 1.0.0        |
| cocoapods-search          | 1.0.0        |
| cocoapods-stats           | 1.0.0        |
| netrc                     | 0.11.0       |
| cocoapods-trunk           | 1.3.0        |
| cocoapods-try             | 1.1.0        |
| colored2                  | 3.1.2        |
| escape                    | 0.0.4        |
| fourflusher               | 2.0.1        |
| gh_inspector              | 1.1.3        |
| molinillo                 | 0.5.7        |
| ruby-macho                | 1.1.0        |
| nanaimo                   | 0.2.5        |
| xcodeproj                 | 1.5.9        |
| cocoapods                 | 1.3.1        |
| colored                   | 1.2          |
| highline                  | 1.7.10       |
| commander-fastlane        | 4.4.6        |
| declarative               | 0.0.10       |
| declarative-option        | 0.1.0        |
| unf_ext                   | 0.0.7.5      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20180417 |
| dotenv                    | 2.5.0        |
| emoji_regex               | 0.1.1        |
| excon                     | 0.62.0       |
| multipart-post            | 2.0.0        |
| faraday                   | 0.15.2       |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| faraday_middleware        | 0.12.2       |
| fastimage                 | 2.1.3        |
| jwt                       | 2.1.0        |
| little-plugger            | 1.1.4        |
| multi_json                | 1.13.1       |
| logging                   | 2.2.2        |
| memoist                   | 0.16.0       |
| os                        | 0.9.6        |
| signet                    | 0.8.1        |
| googleauth                | 0.6.2        |
| httpclient                | 2.8.3        |
| mime-types-data           | 3.2016.0521  |
| mime-types                | 3.1          |
| uber                      | 0.1.0        |
| representable             | 3.0.4        |
| retriable                 | 3.1.2        |
| google-api-client         | 0.21.2       |
| json                      | 2.1.0        |
| mini_magick               | 4.5.1        |
| multi_xml                 | 0.6.0        |
| plist                     | 3.4.0        |
| rubyzip                   | 1.2.1        |
| security                  | 0.1.3        |
| naturally                 | 2.2.0        |
| simctl                    | 1.6.5        |
| slack-notifier            | 2.3.2        |
| terminal-notifier         | 1.8.0        |
| unicode-display_width     | 1.4.0        |
| terminal-table            | 1.8.0        |
| tty-screen                | 0.6.4        |
| tty-cursor                | 0.5.0        |
| tty-spinner               | 0.8.0        |
| word_wrap                 | 1.0.0        |
| rouge                     | 2.0.7        |
| xcpretty                  | 0.2.8        |
| xcpretty-travis-formatter | 1.0.0        |
| fastlane-plugin-sentry    | 1.5.0        |
| oauth                     | 0.5.4        |
| jira-ruby                 | 1.5.0        |
</details>

**Comment by DavidWiesner** [comment]
I had the same issue often in alternation with this issue #12704

**Comment by elviin** [comment]
I also have observed that **pilot's** changelog is not updated when this error occurs but the build is in Testflight uploaded and processed.

**Comment by elviin** [comment]
It looks like the users of the Testflight app also can see that error, so it comes from Apple's API:
https://www.reddit.com/r/TheSilphRoad/comments/8ux7lx/app_download_error_ios/

**Comment by alexs555** [comment]
we also have to open tab Testflight in ITConnect web-interface in order to get new build in Testflight mobile app in iOS

**Comment by elviin** [comment]
@alexs555 Please report it to Apple. There are some people having issues with the user notifications: the test users are not notified before the developer visits the testflight tab in the browser. Really strange.

**Comment by alexs555** [comment]
@elviin I already did it.

Engineering has determined that your bug report is a duplicate of another issue and will be closed.   The open or closed status of the original report your bug was duplicated to appears in a text box within the bug detail section of the bug reporter user interface. For security and privacy reasons, we don't provide access to the original bug yours was duped to. 

Actually, the original bug is still open

**Comment by jamesmontemagno** [comment]
I am experiencing this for Spaceship. Attempting to use the clean_testflight_testers. I assume it is having issues getting all testflight users.

**Comment by elviin** [comment]
@jamesmontemagno Please report it to Apple. Otherwise they will not pay enough attention to this issue. I'm struggling to get from Apple the information why this message comes from their app (Testflight). They do not acknowledge that because I use tools like Fastlane, Circle.

**Comment by jamesmontemagno** [comment]
I tried again this morning and was able to run my query, however when I tried to run it for a second time I got the same thing :( will have to wait until tomorrow.

**Comment by elviin** [comment]
@jamesmontemagno To me it looks like random event. Just try it for the third/fourth time. How many builds do you have in your version in the Testflight uploaded? How many appstore app profiles are you managing with the apple id account?

**Comment by jamesmontemagno** [comment]
We have a few dozen builds. We have about 9000 beta testers though and that is where it was getting caught up at.

**Comment by elviin** [comment]
@jamesmontemagno I am trying to understand the reasons for the rate limits. If that is related to the number of testers/builds/managed accounts.

**Comment by jamesmontemagno** [comment]
It was when I called:  Spaceship::TestFlight::Tester.all(app_id: app_id)

**Comment by elviin** [comment]
I have made a workaround for that, but it is not a proper solution:

I have added the error handling:

begin
 pilot(...)
rescue
    rescue => pilotError

      # code that deals with some exception
      print "[ERROR]: #{pilotError.message}\n"
      print pilotError.backtrace.join("\n\t")

      # The following error symptoms are not breaking the upload to the Tesftlight
      errorSymptoms = ["HTTP ERROR: 403", "Error 403", "Rate limit has been exceeded for:", "description length:"]
      
      # Does the message contain any of the symptoms?
      selected = errorSymptoms.select do |symptom| pilotError.message.include?(symptom) end
      
      if(selected.count.positive?)
        # The error is not breaking the build. Notify the testers and try to upload 
        # the symbols to sentry.

        uploadSymbols(application_debug, 'latest', SENTRY_API_HOST, SENTRY_AUTH_TOKEN, SENTRY_ORG_SLUG, SENTRY_PROJECT_SLUG, BRAND) 
        reportToJira(jira_task_issue_id, scheme_name, application_debug, build_number_string)
      end

      # The error is severe. Do not try to upload the symbols or to report it
      # to the testers.
      
    else
    end

**Comment by keith** [comment]
Still an issue


--
Keith Smiley

**Comment by bpollock-vida** [comment]
Experienced this today, build did not upload to iTunes connect.

Fastlane version: 2.79.0

/Users/travis/.rvm/gems/ruby-2.4.3/gems/fastlane-2.79.0/spaceship/lib/spaceship/test_flight/client.rb:287:in `handle_response': [31m[!] Rate limit has been exceeded for: testflight|default|APPLE_CONNECT:10464547290 (Spaceship::UnexpectedResponse)

**Comment by keith** [comment]
Saw this today with 2.104.0 as well

**Comment by keith** [comment]
Still getting this

**Comment by bpollock-vida** [comment]
Failed due to this reason just today:
Rate limit has been exceeded for: testflight|default|APPLE_CONNECT[0m

    https://github.com/fastlane/fastlane/issues/12845

**Comment by bpollock-vida** [comment]
ring, ring, is this thing on? I get this like every day when I'm trying to deploy a build to apple.

**Comment by bpollock-vida** [comment]
Unable to debug issues locally, because every time I run fastlane (2.107.0) I run into this error:

[09:56:28]: All required keys, certificates and provisioning profiles are installed 🙌
[09:56:28]: Setting Provisioning Profile type to 'app-store'
[09:56:29]: -------------------
[09:56:29]: --- Step: is_ci ---
[09:56:29]: -------------------
+------------------------------------+---------------------------------------------------+
|                                      Lane Context                                      |
+------------------------------------+---------------------------------------------------+
| DEFAULT_PLATFORM                   | ios                                               |
| PLATFORM_NAME                      | ios                                               |
| LANE_NAME                          | ios deploy                                        |
| SIGH_PROFILE_TYPE                  | app-store                                         |
| MATCH_PROVISIONING_PROFILE_MAPPING | {"<redacted>"=>"match AppStore <redacted>"} |
+------------------------------------+---------------------------------------------------+
[09:56:29]: Rate limit has been exceeded for: testflight|default|APPLE_CONNECT:10464547290


+------+----------------------------+-------------+
|                fastlane summary                 |
+------+----------------------------+-------------+
| Step | Action                     | Time (in s) |
+------+----------------------------+-------------+
| 1    | Verifying fastlane version | 0           |
| 2    | default_platform           | 0           |
| 3    | is_ci                      | 0           |
| 4    | match                      | 2           |
| 5    | is_ci                      | 0           |
+------+----------------------------+-------------+

[09:56:29]: fastlane finished with errors

Looking for related GitHub issues on fastlane/fastlane...

➡️  Rate limit has been exceeded for: testflight|default|APPLE_CONNECT
    https://github.com/fastlane/fastlane/issues/12845 [open] 23 💬
    15 hours ago

🔗  You can ⌘ + double-click on links to open them directly in your browser.
/Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/spaceship/lib/spaceship/test_flight/client.rb:330:in `handle_response': [!] Rate limit has been exceeded for: testflight|default|APPLE_CONNECT:10464547290 (Spaceship::UnexpectedResponse)

	from /Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/spaceship/lib/spaceship/test_flight/client.rb:31:in `get_build_trains'
	from /Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/spaceship/lib/spaceship/test_flight/build_trains.rb:16:in `all'
	from /Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/spaceship/lib/spaceship/test_flight/build.rb:108:in `all'
	from /Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/spaceship/lib/spaceship/test_flight/build.rb:119:in `all_processing_builds'
	from /Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/spaceship/lib/spaceship/tunes/application.rb:321:in `all_processing_builds'
	from Fastfile:81:in `fetch_latest_build_version'
	from Fastfile:97:in `next_version_and_build_number'
	from Fastfile:56:in `block (2 levels) in parsing_binding'
	from /Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/fastlane/lib/fastlane/lane.rb:33:in `call'
	from /Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/fastlane/lib/fastlane/runner.rb:49:in `block in execute'
	from /Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/fastlane/lib/fastlane/runner.rb:45:in `chdir'
	from /Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/fastlane/lib/fastlane/runner.rb:45:in `execute'
	from /Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/fastlane/lib/fastlane/lane_manager.rb:56:in `cruise_lane'
	from /Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/fastlane/lib/fastlane/command_line_handler.rb:36:in `handle'
	from /Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/fastlane/lib/fastlane/commands_generator.rb:108:in `block (2 levels) in run'
	from /Library/Ruby/Gems/2.3.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:178:in `call'
	from /Library/Ruby/Gems/2.3.0/gems/commander-fastlane-4.4.6/lib/commander/command.rb:153:in `run'
	from /Library/Ruby/Gems/2.3.0/gems/commander-fastlane-4.4.6/lib/commander/runner.rb:476:in `run_active_command'
	from /Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:76:in `run!'
	from /Library/Ruby/Gems/2.3.0/gems/commander-fastlane-4.4.6/lib/commander/delegates.rb:15:in `run!'
	from /Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/fastlane/lib/fastlane/commands_generator.rb:333:in `run'
	from /Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/fastlane/lib/fastlane/commands_generator.rb:41:in `start'
	from /Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/fastlane/lib/fastlane/cli_tools_distributor.rb:108:in `take_off'
	from /Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/bin/fastlane:23:in `<top (required)>'
	from /Users/bricepollock/.gem/ruby/2.3.0/bin/fastlane:22:in `load'
	from /Users/bricepollock/.gem/ruby/2.3.0/bin/fastlane:22:in `<main>'

**Comment by DavidWiesner** [comment]
Still an issue in latest version

**Comment by janpio** [comment]
As @elviin said above, this error is coming from Apple's server in a pretty unexpected way as there doesn't seem to be any specific reason to hit the rate limiting. So there is no real way to fix this (yet).

---

One thing I am not clear about yet, what exactly are the negative results of this happening? @elviin mentioned the testflight release notes not being uploaded, anything else?

@bpollock-vida What is the `Fastfile` being executed in https://github.com/fastlane/fastlane/issues/12845#issuecomment-433474967? It seems `is_ci` is successful, and then something causes the testflight rate limiting which makes no real sense as there should be no additional action executed, correct?

---

In the background `spaceship`, our library for communication with Apple's servers, is handling the requests to Apple's servers, including the ones for Testflight. 

It would be helpful if you could check your [spaceship logfiles](https://github.com/fastlane/fastlane/blob/master/spaceship/docs/Debugging.md#logging) or even [create a Charles session](https://github.com/fastlane/fastlane/blob/master/spaceship/docs/Debugging.md#proxy-support) of the traffic when this occurs (Send them to piotrowski@gmail.com and mention this issue if you don't just want to post them publicly here, which makes a lot of sense...)

If we collect some of those log files or Charles sessions, we will be able to understand better what spaceship is doing _before_ you all get this error. This could enable us to maybe identify the culprit and fix it - and at least allow us to implement a better error message.

**Comment by bpollock-vida** [comment]
Hey @janpio I'm glad someone has finally responded. 

The affect is my scheduled daily builds on travis can fail 50% of the days of the week some weeks. It also means for a period of time that can exceed 30min I may be unable to deploy with my current script (even locally) and will have to modify my fast file to ignore the offending fastlane operations to unblock.

There is one simple solution that would be to try to implement a retry and backoff, although if this happens regularly over a long window that might not be effective.

The offender that hits the rate limiting is at the top of the backtrace in 'all_processing_builds'.
/Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/spaceship/lib/spaceship/test_flight/client.rb:330:in `handle_response': [!] Rate limit has been exceeded for: testflight|default|APPLE_CONNECT:10464547290 (Spaceship::UnexpectedResponse)

	from /Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/spaceship/lib/spaceship/test_flight/client.rb:31:in `get_build_trains'
	from /Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/spaceship/lib/spaceship/test_flight/build_trains.rb:16:in `all'
	from /Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/spaceship/lib/spaceship/test_flight/build.rb:108:in `all'
	from /Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/spaceship/lib/spaceship/test_flight/build.rb:119:in `all_processing_builds'
	from /Library/Ruby/Gems/2.3.0/gems/fastlane-2.107.0/spaceship/lib/spaceship/tunes/application.rb:321:in `all_processing_builds'
	from Fastfile:81:in `fetch_latest_build_version'
	from Fastfile:97:in `next_version_and_build_number'

Code:
```
  def fetch_latest_build_version
    Spaceship::Tunes.login("<redacted>")
    app = Spaceship::Tunes::Application.find('<redacted>')
    processing_builds = app.all_processing_builds
    processed_builds = app.builds
    most_recent_build = (processing_builds + processed_builds).sort_by {|build| build.upload_date}.last
    latest_version = Versionomy.parse(most_recent_build.train_version)
    latest_build_number = latest_testflight_build_number(version: latest_version.to_s)

    return latest_version, latest_build_number
  end
```

**Comment by janpio** [comment]
@bpollock-vida So this `fetch_latest_build_version` is executed after `is_ci` and then failing? What are you doing with these 2 numbers so it can fail a build? What is the significance of this "a period of time that can exceed 30min"? Is access blocked so long? 

---

> There is one simple solution that would be to try to implement a retry and backoff, although if this happens regularly over a long window that might not be effective.

Does anyone know how long the rate limiting is in effect? Because we could catch this response from Testflight and handle it - but the question is how. If the effect stays for ~30 minutes, a retry of course won't work.

Anyone know the significance of the value after `APPLE_CONNECT:`?

The most useful thing here really would be spaceship logs or a Charles session, so we can really understand what is _causing_ the rate limiting: https://github.com/fastlane/fastlane/issues/12845#issuecomment-444164180

**Comment by bpollock-vida** [comment]
_So this fetch_latest_build_version is executed after is_ci and then failing?_
is_ci is the lane we use for our travis deploys. On execution, we call the fetch_latest_build_version function to get prior build information so we can increment the build and/or version. Apple does not allow duplicate uploads of the same build-version combination so this is a requirement for automatic builds.

_What are you doing with these 2 numbers so it can fail a build?_
If you look at the stack trace and code, it is the 'all_processing_builds' method on the fastlane supplied spaceship 'app' that throws an error that fails the build. It does not appear to be my specific Fastfile.

_What is the significance of this "a period of time that can exceed 30min"? Is access blocked so long?_
This is my anecdotal description of the behavior I've observed. Yes, access can be blocked for a long time.

_Because we could catch this response from Testflight and handle it_
Definitely one thing that can be done is provide better logging in or from Spaceship so we get an actionable error message with more information.

**Comment by janpio** [comment]
I know what actual code is crashing. 

But the method `all_processing_builds` mentioned there one only appears because the "hey, you can't make more requests because you triggered the rate limiting" response does not output a nice error message yet. I can add this handling in a few minutes - but the only thing I can make it say is "no idea why this happened, but you are rate limited now. try again in half an hour or so.". And that is not helpful at all. So I won't do that before we don't have more information.

That is why I want to know what your lane is doing _before_ it fails for this request. Something has to make your account go over this rate limit on Apple's server. 

Spaceship logging with 2.109.x should be pretty extensive, but a Charles Web Proxy session would still be better as it contains a lot more information (logging compete http requests to a flat file unfortunately doesn't scale very well),

**Comment by bpollock-vida** [comment]
Attached the relevant code, however, let me provide a summary:
* travis calls deploy lane
* fastlane executes match against 'appstore'
* we try to fetch the latest build version from spaceship
* we hit the crash

Since there is very little happening before the spaceship call, it must be match that is querying so much to produce the rate limiting?

We run the 'test' lane on every build, one other thought is if 'scan' somehow is hitting apple servers that would build up a lot more requests than match happening once daily before the build version request. However, I don't see any evidence that is happening.

Test code:
```
  desc "Runs all the tests"
  lane :test do
    scan(
	  workspace: "<redacted>.xcworkspace",
	  scheme: "<redacted>",
	  devices: ["iPhone 6"],
	  clean: true,
	  formatter: "./fastlane/xcpretty_formatter.rb"
    )
  end
```

Deploy Code:
```
# travis calls...
fastlane deploy 

# ...our deploy lane:
  desc "Build and deploy to both Testflight and Crashlytics"
  desc "Also ensures an up to date provisioning profile for crashlytics"
  lane :deploy do |options|
    version, build_number = next_version_and_build_number(options)

    deploy_testflight
    deploy_crashlytics

    app_version = "v#{version.to_s}-#{build_number}"
    add_git_tag(tag: app_version)
    push_git_tags

    slack(
      slack_url: "<redacted>",
      message: "Successfully deployed iOS to Testflight and Crashlytics",
      channel: "<redacted>",
      success: true,
      payload: {
        'Version' => "v#{version.to_s} (#{build_number})",
        "Build Date" => Time.new.to_s,
        "Branch" => git_branch
      },
      default_payloads: [:last_git_commit_hash]
    )
  end
  
  # Must be logged in with a match profile
  # NOTE: On travis we hit the else case in this function  
  def next_version_and_build_number(options = nil)
    if options[:version_number]
      next_version = options[:version_number]
      next_build_number = (options[:build_number] ? options[:build_number] : 1)

    else 
      ci_match_helper(type: "appstore")
      next_version, latest_build_number = fetch_latest_build_version
      next_build_number = latest_build_number + 1

      if should_bump_version
        next_version = next_version.bump(:tiny)
        next_build_number = 1
      end
    end

    increment_version_number(version_number: next_version.to_s)
    increment_build_number(build_number: next_build_number)

    return next_version, next_build_number
  end
  
  def ci_match_helper(**params)
    if is_ci
      match(**params, keychain_name: <redacted>)
    end

    match(**params)
  end
  
  def fetch_latest_build_version
    Spaceship::Tunes.login("<redacted>")
    app = Spaceship::Tunes::Application.find('<redacted>')
    processing_builds = app.all_processing_builds
    processed_builds = app.builds
    most_recent_build = (processing_builds + processed_builds).sort_by {|build| build.upload_date}.last
    latest_version = Versionomy.parse(most_recent_build.train_version)
    latest_build_number = latest_testflight_build_number(version: latest_version.to_s)

    return latest_version, latest_build_number
  end
```

**Comment by janpio** [comment]
Thanks. That really doesn't look like anything should trigger any relevant requests here. What is `match` doing according to your output? Usually it should be happy with the certificates from the git repo and just install those, right?

`scan` doesn't hit any APIs at all, and `match` talks to the Apple Developer Portal, not App Store Connect where Testflight lives - so different APIs.

When you hit the rate limiting, you can still run `match` afterwards, right? 
Only commands that are Testflight related will trigger the error, correct?

We will need to see some spaceship logs to be able to understand what is going on here: https://github.com/fastlane/fastlane/issues/12845#issuecomment-444164180 Both for understanding what is causing this and implementing proper error handling.

**Comment by bpollock-vida** [comment]
Looking at the logging I captured above, it looks like match says everything is cool:
[09:56:28]: All required keys, certificates and provisioning profiles are installed 🙌
[09:56:28]: Setting Provisioning Profile type to 'app-store'

_When you hit the rate limiting, you can still run match afterwards, right?
Only commands that are Testflight related will trigger the error, correct?_
Yeah its just the getting the build version details where we fail every time.

We are trying to set it up so we can capture these spaceship logs more easily when we see this issue. Will update when we capture them. Unless one of the many others experiencing this issue can help and post theirs. :)

**Comment by bpollock-vida** [comment]
We were able to get spaceship logs today and pursued some investigation. We ruled out that our 'sort_by' function was the causing the problem and narrowed it down to #all_processing_builds and #builds. 

There appears to be several challenges:
* We have many trains and these functions make one query request per train
* Both functions make a query request for every train.
   - #all_processing_builds filters the result to provide just the processing builds
   - #builds filters the result to provide non-processing builds
   - there is no spaceship function that provides all builds without filtering
* If we just call one of these, we won't timeout. However, every time we call both we get a timeout. (maybe the rate limit is between N and 2N?)

Spaceship solution:
 1. Provide a function that provides all builds without filtering
 2. Use some sort of caching to avoid the duplicated requests for each train

A work around:
 1. In the Fastfile don't use build numbers only trains. Use the app.trains call, then convert train strings to a number to determine last published train.
(results in 1 network request)
2. In the Fastfile, get all trains and query for build information in the Fastfile for each to find 'last upload date'. 
(results in N network requests which seems to be below the rate limiting)

[spaceship logs.zip](https://github.com/fastlane/fastlane/files/2655044/spaceship.logs.zip)

**Comment by janpio** [comment]
Going from your erroring log, we are talking about ~140 requests in ~2 minutes. Hm, that is certainly a lot, but also not a DDoS. Let's see what all the other people in this issue will report back.

---

Now to your specific problem:
Unless I am mistaken `app.all_processing_builds` is a just a filtered subset of `app.builds`:
https://github.com/fastlane/fastlane/blob/cd88359fa9b6772f91a31a290f8f3c790a716dc0/spaceship/lib/spaceship/tunes/application.rb#L320-L322
https://github.com/fastlane/fastlane/blob/cd88359fa9b6772f91a31a290f8f3c790a716dc0/spaceship/lib/spaceship/tunes/application.rb#L350-L354
https://github.com/fastlane/fastlane/blob/cd88359fa9b6772f91a31a290f8f3c790a716dc0/spaceship/lib/spaceship/test_flight/build.rb#L118-L120
https://github.com/fastlane/fastlane/blob/cd88359fa9b6772f91a31a290f8f3c790a716dc0/spaceship/lib/spaceship/test_flight/build.rb#L107-L110

Also, `latest` (used as `Spaceship::TestFlight::Build.latest(app_id: app_id, platform: platform)`) might be what you actually want in your `Fastfile`:
https://github.com/fastlane/fastlane/blob/cd88359fa9b6772f91a31a290f8f3c790a716dc0/spaceship/lib/spaceship/test_flight/build.rb#L127-L129

(But again, this is just from a short look - we should possibly split this discussion in a new issue if we want to continue that.)

---

Better question: How should we handle this timeout? Now, with a log in hand, I created #13841 to implement proper error handling. But right now I don't really know what we could do about it. Please go to the issue #13841 and post your suggestions!

**Comment by bpollock-vida** [comment]
Does look like builds returns all, going to give that a shot to move from 2N to N calls. I want to stay in iTunesConnect not TestFlight so I'll avoid Spaceship::TestFlight::Build.latest for now.

**Comment by bpollock-vida** [comment]
Is there a way I can catch this error from spaceship? being-rescue doesn't seem to work because Spaceship is another process?

**Comment by janpio** [comment]
Feel free to submit a PR to spaceship documentation.

Spaceship is normal ruby code used like everything else. Nothing special in there (afaik).

**Comment by bpollock-vida** [comment]
Ah, here is the documentation that says it is all processed builds:
https://www.rubydoc.info/gems/spaceship/Spaceship/Tunes/Application

#builds(platform: nil) ⇒ Object
Get all builds that are already processed for all build trains You can either use the return value (array) or pass a block.

**Comment by bpollock-vida** [comment]
Is there any documentation for spaceship anymore?

Can we expose Spaceship::TestFlight::Client.get_build_trains(<id>)? The only trains calls that exist today make the 1 request to get trains and then N requests for the build info on each train

**Comment by janpio** [comment]
https://github.com/fastlane/fastlane/tree/master/spaceship + https://github.com/fastlane/fastlane/tree/master/spaceship/docs

For your other question, please best open a new issue. I don't want to spam all those poor participants here with further discussion about this - but it is a good thing to solve.

**Comment by bpollock-vida** [comment]
I don't think this is spamming. There has been no solution for this issue yet in this thread. 

The only provided methods (all or latest) make 1 request for all trains and then 1 request for each train (N) which can still lead to rate limiting if you run fast lane too quickly (even ten minutes apart). 

The only true solution I've seen yet is to expose the spaceship function to get all trains (Spaceship::TestFlight::Client.get_build_trains()) without also getting build info for each train (N).

**Comment by janpio** [comment]
Not everybody here using the same spaceship methods as you are. The discussion here is about how to handle the `testflight|default|APPLE_CONNECT` problem in general. One of the underlying problems might be that some methods create a lot of requests and could be built in a different way - if that is the case, please create a new issue for this specific problem in the code.

**Comment by DavidWiesner** [comment]
There is still no solution in latest version. Still getting this error

**Comment by bpollock-vida** [comment]
@DavidWiesner can you provide the same details I have by capturing spaceship logs or code. That way we can see if the same fast lane code is generating this issue for you. My work around was to only call one builds method to reduce the call overhead.

```i.e. just calling
    processing_builds = app.all_processing_builds
    processed_builds = app.builds

whereas before I called
    processing_builds = app.all_processing_builds
    processed_builds = app.builds```

**Comment by janpio** [comment]
@scorpion35 You are the first person to report the `Rate limit has been exceeded for: itunes-connect|...`  error message here - so this is something different than what the original issue is about. Please create a new issue. Thank you.

**Comment by scorpion35** [comment]
@janpio Sure! Here it is - https://github.com/fastlane/fastlane/issues/14070

**Comment by janpio** [comment]
Thanks for the investigation @mevinDhun!

Could you please open an additional issue for this specific problem? I want to debug this together with you, but as this will require quite a bit of back and forth I would prefer to do that in an issue without so many people getting notified on each comment ;) 
It would be great if you could include the whole spaceship log from `/tmp/spaceship-*.log` of one of those cases where you get the timeout, and the whole output you are getting on the command line. There _has_ to be a way to structure the code better to create less request - we probably "just" have to dive deep in.

**Comment by Sega-Zero** [comment]
Having this error too. Any news on resolving this issue?

**Comment by mevinDhun** [comment]
@Sega-Zero  I managed to fix this issue by setting _wait_for_uploaded_build_:
```ruby
upload_to_testflight(
      ipa: 'platforms/ios/build/device/testapp.ipa',
      wait_for_uploaded_build: true
 )
```

**Comment by Sega-Zero** [comment]
I don't think you've fixed it. I caught that with `wait_for_uploaded_build: true` already ;)

**Comment by DavidWiesner** [comment]
Still got this error in latest release

**Comment by Sega-Zero** [comment]
Also experiencing this today

**Comment by bpollock-vida** [comment]
BTW. Since no one at fastlane is taking this issue seriously. I implemented my own versioning through a google cloud 

```
# We've been running into issues for a year with using fastlane to fetch builds from apple and hitting rate limiting
# due to N numbers of requests where N = number of versions uploaded to apple
#
# Instead we will manage our version/build through a file housed on Google Cloud Services (GCS):
# https://console.cloud.google.com/storage/browser/<bucket_name>?project=<project_id>&organizationId=<org_id>
#
# Documentation on the GCS ruby gem we use:
# https://cloud.google.com/storage/docs/downloading-objects#storage-download-object-ruby
#
# A manual build will ignore these settings, but still update the file at a later date
# This cloud file will be updated whether ran in travis or locally.
require "google/cloud/storage"
require "json"

GCS_REMOTE_FILE_NAME = "<file>.json"
GCS_LOCAL_TEMP_FILE_NAME = "downloaded-version.json"
GCS_PROJECT_ID = "<project_id>"
GCS_BUCKET_NAME = "<bucket_name>"

platform :ios do
  ...
  
  desc "Test out generating the next version"
  lane :check_version do
    version, build = next_version_and_build_number({})
    update_last_version(version, build)
    print "\n[LOG] DEPLOYED version: ", version, " build: ", build, "\n"
  end

  def gcs_download_remote_file()
    storage = Google::Cloud::Storage.new project_id: GCS_PROJECT_ID
    bucket  = storage.bucket GCS_BUCKET_NAME
    file    = bucket.file GCS_REMOTE_FILE_NAME

    file.download GCS_LOCAL_TEMP_FILE_NAME

    puts "Downloaded #{file.name}"
  end

  def gcs_build_version()
    gcs_download_remote_file()

    data = File.read(GCS_LOCAL_TEMP_FILE_NAME)
    json = JSON.parse(data)
    version = json["version"]
    build = json["build"]
    return version, build
  end

  def update_last_version(next_version, next_build)
    gcs_download_remote_file()

    data = File.read(GCS_LOCAL_TEMP_FILE_NAME)
    json = JSON.parse(data)

    json["version"] = next_version
    json["build"] = next_build

    str_json = json.to_json
    output = File.open(GCS_LOCAL_TEMP_FILE_NAME, "w")
    output.write(str_json)
    output.close

    storage = Google::Cloud::Storage.new project_id: GCS_PROJECT_ID
    bucket  = storage.bucket GCS_BUCKET_NAME
    updated_file = bucket.create_file GCS_LOCAL_TEMP_FILE_NAME, GCS_REMOTE_FILE_NAME

    puts "Uploaded #{updated_file.name}"
  end

  # Must be logged in with a match profile
  def next_version_and_build_number(options = {})
    if options[:version_number]
      next_version = options[:version_number]
      next_build_number = (options[:build_number] ? options[:build_number] : 1)
    else
      version, build = gcs_build_version
      print "\n[LOG] EXISTING version: ", version, " build: ", build, "\n"
      next_version = Versionomy.parse(version).bump(:tiny).to_s
      next_build_number = 1
    end

    increment_version_number(version_number: next_version)
    increment_build_number(build_number: next_build_number)

    return next_version, next_build_number
  end
end
```
